src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+ *                              (non-target) in a domain has been used.
+ * \param[in]   used_targets    A list of the targets that have been used. We
+ *                              iterate through this when selecting the next
+ *                              target in a placement to determine if that
+ *                              location is valid.
+ * \param[in]   shard_num       the current shard number. This is used when
+ *                              selecting a target to determine if repeated
+ *                              targets are allowed in the case that there
+ *                              are more shards than targets
+ *
+ * \return      an int value indicating if the returned target is available (0)
+ *              or failed (1)
+ */
+static void
+get_target(struct pool_domain *curr_dom, struct pool_target **target,
+	   uint64_t obj_key, uint8_t *dom_used, uint32_t *used_targets,
+	   int shard_num)
+{
+	uint8_t         found_target = 0;
+	uint8_t         top = 0;
+	uint32_t        fail_num = 0;
+	uint32_t        selected_dom;
+	uint64_t        root_pos;
+
+	root_pos = (uint64_t)curr_dom;
+
+	do {
+		uint32_t        num_doms;
+		uint64_t        key;
+		uint64_t        curr_pos;
+
+		/* Retrieve number of nodes in this domain */
+		if (curr_dom->do_children == NULL)
+			num_doms = curr_dom->do_target_nr;
+		else
+			num_doms = curr_dom->do_child_nr;
+
+		key = obj_key;
+		/* If choosing target in lowest fault domain level */
+
+		curr_pos = (uint64_t)curr_dom - root_pos;
+		curr_pos = curr_pos / sizeof(struct pool_domain);
+
+		if (curr_dom->do_children == NULL) {
+			uint32_t dom_id;
+			uint32_t        i;
+
+			do {
+				/*
+				 * Must crc key because jump consistent hash
+				 * requires an even distribution or it will
+				 * not work
+				 */
+				key = crc(key, fail_num++);
+
+				/* Get target for shard */
+				selected_dom = jump_consistent_hash(key,
+								    num_doms);
+
+				/* Retrieve actual target using index */
+				*target = &curr_dom->do_targets[selected_dom];
+
+				/* Get target id to check if target used */
+				dom_id = (*target)->ta_comp.co_id;
+
+				/*
+				 * Check to see if this target is valid to use.
+				 * You can reuse targets as long as there are
+				 * fewer targets than shards and all targets
+				 * have already been used
+				 */
+				for (i = 0; used_targets[i] != 0; ++i) {
+					if (used_targets[i] == dom_id + 1)
+						break;
+				}
+			} while (used_targets[i]);
+
+			/* Mark target as used */
+			used_targets[i] = dom_id + 1;
+
+			/* Found target (which may be available or not) */
+			found_target = 1;
+		} else {
+
+			int             range_set;
+			uint64_t        child_pos;
+
+			child_pos = (uint64_t)(curr_dom->do_children)
+				    - root_pos;
+			child_pos = child_pos / sizeof(struct pool_domain);
+
+			/*
+			 * If all of the nodes have been used for shards but we
+			 * still have shards to place mark all nodes as unused
+			 * so duplicates can be chosen
+			 */
```

Comment:
So two replicas could be located to same domain? Should we check the object class (If it's replication or EC and what's the replication domain level (node by default so far)) and decide if it's allowed to put two shards in same domain?

Reply:
This was an old comment, This is no longer true and placement will return an error if attempting to place an object who's sharding requirements cannot be met.

removed.

Reply:
Yes I just need to remove it in the next push.

The sharding requirements will be checked during "mapless_obj_spec_place_get(...)" or more specifically "spec_place_rank_get(...)" if there are not enough domains to fulfill the requirements  then it returns.

Reply:
In this case the shards should be placed as far apart as possible and will only be placed in the same domain once all others have been used. This goes for any domain level. 

This should function in the same way that the ring map does.
So objects are only co-located if the object class is larger than the number of targets. 

If we don't have enough targets for an object should we just return failure?

Reply:
For now, we just set po_target -1 to let the shard unavailable instead of return failure.
Hmm, I am not sure we want to put multiple shards in the same domain at least for now. especially each domain is just 1 node(rank) in our test.

Reply:
Let's assume user want to create a 3-way replica object, but it failed to find enough fault domains, I think we'd simply return failure instead of returning a 2-way replica silently.

The placement map unit tests (src/placement/tests) should be improved to test all these cases.

Reply:
So lets say we have 3 servers, with 1 node each, and each node has 8 targets, 

If we try and place any object with more than 3 shards total should we return failure?

Reply:
That depends, if user's expectation is to create node-level replicas, more than 3 shards creation should fail, if user is to create a striped object, I think it's ok to locate multiple shared on same node.

Anyway, I think we can refine this and improve unit tests in follow-on patches.